ast/parser.go

```diff
@@ -719,67 +737,65 @@ func (p *Parser) parseElse(head *Head) *Rule {
 
 func (p *Parser) parseHead(defaultRule bool) (*Head, bool) {
 
-	var head Head
-	head.SetLoc(p.s.Loc())
-
+	head := &Head{}
+	loc := p.s.Loc()
 	defer func() {
-		head.Location.Text = p.s.Text(head.Location.Offset, p.s.lastEnd)
+		if head != nil {
+			head.SetLoc(loc)
+			head.Location.Text = p.s.Text(head.Location.Offset, p.s.lastEnd)
+		}
 	}()
 
-	if term := p.parseVar(); term != nil {
-		head.Name = term.Value.(Var)
-	} else {
-		p.illegal("expected rule head name")
+	term := p.parseVar()
+	if term == nil {
+		return nil, false
 	}
 
-	p.scan()
-
-	if p.s.tok == tokens.LParen {
-		p.scan()
-		if p.s.tok != tokens.RParen {
-			head.Args = p.parseTermList(tokens.RParen, nil)
-			if head.Args == nil {
-				return nil, false
-			}
+	ref := p.parseTermFinish(term, true)
```

Comment:
So we parse it into a Ref and interpret the result -- below 👇 

This allows us to de-duplicate the parser logic for head refs and other "normal" refs.